### ISSUE TITLE
Communication thread pinning on CoolMUC2

### ISSUE THREAD

**Issue by psamfass** [chameleon-hpc/chameleon#12]
The communication thread pinning on CoolMUC2 (LRZ Linux Cluster) does presently not work correctly: the thread is pinned to core id 26 instead of core id 27. It seems as if CPU_ISSET does not recognize that the logical core id 55 is in the CPU set.

**Comment by jklinkenberg** [comment]
The question here is whether that is a problem with CoolMUC2 or the environment there or Chameleon.
On other partitions / clusters it works fine.

**Comment by jklinkenberg** [comment]
Solution in commit: 51f38e6
Fixed bug or interplay between Chameleon and OpenMP runtime that already performed thread binding before process cpuset is read. Now getting process cpuset happens before thread binding occurs, at least for Intel and Clang.

For GNU there is still a blocking compiler issue: #15

**Comment by uphoffc** [comment]
Hi Jannis,

I had some problems recently with GOMP and thread binding. Our solution was to just MPI_Allreduce the cpu masks with a shared node split communicator. Works pretty well imho, so maybe this is helpful:

https://github.com/SeisSol/SeisSol/blob/master/src/Parallel/Pin.cpp#L143

Regards,
Carsten

**Comment by jklinkenberg** [comment]
Hi Carsten,

I will have a look into it but I guess the problem is much more complicated.
Consider a scenario with 2 or 4 processes per node. If you use the GNU compiler and specify an affinity policy like `OMP_PLACES=cores OMP_PROC_BIND=spread` within each process then the binding of the threads (also for the master thread of a process) happens even before the main functions is executed. Thus, the master threads is already restricted to a single core and it is pretty hard to get the initial cpuset for the whole process.

I experimented a lot with that in the last days and hopefully found a solution.
However, currently it is hard to test the GNU compiler with Chameleon as we have another compilation problem.
For Intel and Clang that is working.

Best regards,
Jannis

**Comment by uphoffc** [comment]
Hi Jannis,

yes I'm well aware of that problem. For Intel you have to get the process mask before you do anything else, but for GCC its hopeless....

My idea was that if you cannot get the process mask in a reliable way, then one shouldn't even attempt to read it :-D But the following workaround works pretty well:
1. First you get the CPU mask covered by the OpenMP threads by running sched_getaffinity in each thread (getWorkerUnionMask)
2. You split the MPI communicator with MPI_COMM_TYPE_SHARED and then you Allreduce the CPU mask  on each node (getNodeMask)
3. Now you know the CPU mask covered by OpenMP threads on the node, i.e. that works also for multiple ranks per node. You can now distribute the free cores to communication threads, either in a round-robin fashion (i.e. by using the local rank of the split communicator) or you use libnuma to figure out the socket affinity of the free cores (or a combination of both).

Regards,
Carsten

**Comment by jklinkenberg** [comment]
Hi Carsten,

I actually found a solution that also works with the GNU compiler.
The problem is that GNU's libgomp does its initialization before main is executed in a section with `__attribute__((constructor)`. It is pretty hard to execute any code before that. However, if you are not compiling a shared library but an executable you are able to annotate a code section with `__attribute__((section(".preinit_array")))` which runs even before the libgomp constructors.

For Chameleon it is a bit more tricky, as it is compiled as shared library and `.preinit_array` cannot be used in DSOs.
Solution: I created a new header file that can be compiled into the application that passes the cpuset to Chameleon before initialization.

If you want to see details, I implemented that in the following commits:
chameleon: [f813964](https://github.com/chameleon-hpc/chameleon/commit/f813964d90ae146d2cb3c22c0c1e2cf9f8e0ff44)
chameleon-apps: [7e88ff4](https://github.com/chameleon-hpc/chameleon-apps/commit/7e88ff4b8639e8d52a3d973352edb27330941f27)

That works for me and the approach should work for any C/C++ application independent of MPI.

Best,
Jannis
